### PR TITLE
KAFKA-8637: WriteBatch objects leak off-heap memory

### PR DESCRIPTION
**Pull request by ableegoldman**
Should be cherry-picked back to 2.3 (picked from 2.2 to 2.1 in  [7077](https://github.com/apache/kafka/pull/7077) )

### CODE REVIEW

**Comment by ableegoldman** [comment]
@bbejeck @mjsax @vvcephei @cadonna @abbccdda @guozhangwang

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java [hunk 1]
```diff
@@ -403,10 +403,10 @@ public synchronized void close() {
         }
 
         dbAccessor.close();
+        db.close();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Can we add a comment and maybe link to RocksDB docs to clarify the correct order of close() calls? Otherwise, we might forget about it and reorder accidentally...

Also: this may affect more versions compared to the memory leak? Should we extract it into it's own PR to easily backport?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStore.java [hunk 2]
```diff
@@ -233,6 +233,7 @@ void restoreAllInternal(final Collection<KeyValue<byte[], byte[]>> records) {
                 final S segment = entry.getKey();
                 final WriteBatch batch = entry.getValue();
                 segment.write(batch);
+                batch.close();
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStore.java:236]: Can we add a test this this?

**Comment by bbejeck** [comment]
retest this please

**Comment by bbejeck** [comment]
retest this please

**Comment by bbejeck** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 1 above)
Comment (pkleindl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Looking again at the docs https://github.com/facebook/rocksdb/wiki/RocksJava-Basics I am still not sure if the order is now completely correct.
The doc says ColumnFamilyHandle > RocksDB > DBOptions > ColumnFamilyOptions
but the code does ColumnFamilyHandle > ColumnFamilyOptions > DBOptions > RocksDB due to userSpecifiedOptions.
Additionally, only some are set to null afterwards, does this have a special reason?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStore.java (hunk 2 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStore.java:236]: Not really sure what a good way to test this would be...thoughts?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 1 above)
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: This PR is moving the `db.close()` to before `userSpecifiedOptions.close()` so aren't we now closing the db in the right order? That said, the docs do suggest we should flip the order in which `DBOptions` & `ColumnFamilyOptions` are closed in `RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter#close` (which, side note, is an obscenely long name...)

But Matthias is right, this should be split into a separate PR (and well documented) -- will open another one

Also, I think on this branch everything *is* set to null -- but if you were looking at the 2.2 branch, yeah I did forget to set filter to null. We were rushing to get that in before 2.2.1 was cut 😄

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStore.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractRocksDBSegmentedBytesStore.java:236]: I would be tricky. We would need to intercept the call to `new WriteBatch` within `getWriteBatches()` using power-mock and later verify if `close()` was called on the mock. Might not be worth it...

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Thanks for splitting out https://github.com/apache/kafka/pull/7076 @ableegoldman 

Please review the new PR, too, @pkleindl

**Review by bbejeck — APPROVED** [review]

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java (hunk 1 above)
Comment (pkleindl) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Will do.
My question regarding the null setting was about the DBOptions & ColumnFamilyOptions in RocksDBGenericOptionsToDbOptionsColumnFamilyOptionsAdapter, why they are not set to null too.
As for the long name, it's probably a german language thing ;-)

**Comment by mjsax** [comment]
Java11 2.12 timed out. Java11 2.13 and Java8 failed with flaky tests.

Retest this please.

**Comment by ableegoldman** [comment]
Java 8 passed, Java11 2.13 failed with flaky tests. retest this, please

**Comment by mjsax** [comment]
Java8 and 11/2.12 passed. 11/2.13 failed (test result not available any longer).

Retest this please.

**Review by guozhangwang — APPROVED** [review]

**Comment by guozhangwang** [comment]
Cherry-picked to 2.3 as well.
